### ISSUE TITLE
The symlink to exec root seems to be named incorrectly

### ISSUE THREAD

**Issue by mrkkrp** [bazelbuild/bazel#5506]
### Description of the problem / feature request:

According to this documentation: https://docs.bazel.build/versions/master/output_directories.html, the symlink to exec root should be named `bazel-<workspace-name>`, but in practice it's named `bazel-<containing-dir-name>`, which may be different.

For example, if I have a `WORKSPACE` file with this:

```
workspace(name = "foo")
```

contained in directory named `bar`, then inside `bar` Bazel creates `bazel-bar`, not `bazel-foo` as it should.

This is inconvenient sometimes because I don't see how to get the name of containing directory (`bar`), while workspace name (`foo`) is available from `ctx`.

### What operating system are you running Bazel on?

NixOS:

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.48, NixOS, 18.09pre142796.4b649a99d84 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1pre6148_a4aac7f`
 - channels(root): `"nixos-18.09pre142796.4b649a99d84"`
warning: Nix search path entry '/home/mark/.nix-defexpr/channels' does not exist, ignoring
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### What's the output of `bazel info release`?

```
release 0.12.0- (@non-git)
```

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

Got it from `nix`.

###  Have you found anything relevant by searching the web?

No.

### Any other information, logs, or outputs that you want to share?

No.

**Comment by lfpino** [comment]
Assigning to @lberki for triage.

**Comment by mrkkrp** [comment]
Well, we allow our users to build REPLs which are shell scripts. These shell scripts are run with `--direct_run` from root of workspace and they need to know locations of various things that were built (shared libraries, etc.). So how to infer where to find those things?

**Comment by lberki** [comment]
`--direct_run` (which is the default and only option now) runs your binaries in their runfiles directory and if they need something, they should declare a data dependency on them, and then the needed files will be in the runfiles directory.

The cwd where you ran Bazel and the workspace root are available as `$BUILD_WORKING_DIRECTORY` and `$BUILD_WORKSPACE_DIRECTORY`, respectively.

**Comment by mrkkrp** [comment]
That's useful information, thanks! Perhaps we indeed can refactor that code now that we have `--direct_run`.

**Comment by blackliner** [comment]
Hi, for me the naming of the bazel-<directory-name> is a problem, because `bazel clean` does only clean this symlink if it has the same name as the parent dir.
For example, if i have  `~/my-project` and run `bazel build ...`, a symlink named `bazel-my-project` is generated.
If i now map `~/my-project` into a docker as `/src/`, bazel goes nuts because it finds files in the `bazel-my-project` symlink it cannot handle. A `bazel clean` does not delete this symlink, only a `bazel-src` if it is there.

**Comment by janakdr** [comment]
I agree that our documentation is wrong here. It's actually wrong in multiple directions: the directory under the execroot should be "my-project", not "workspace-name". However, there's a basic problem with using the repository name versus the workspace name for the convenience symlink: on `bazel clean`, we need to know the repository name, but we can't get that without loading the `WORKSPACE` file. We don't currently have a way to only extract the `name` attribute from the `WORKSPACE` file, so that means that a `clean` could take a long time, as the `WORKSPACE` file is fully processed. That isn't the expected behavior of `bazel clean`. We could write a simplified parser that only looked for `name` (it does have to be at the top), but that seems kind of janky.

cc @mhy1992

**Comment by janakdr** [comment]
@blackliner is it important to change the name of the directory when moving into Docker? In some ways, it would be simpler if we changed behavior to match the documentation, and made all of these subdirectories have the name of the workspace *directory*, rather than the repository name. What is the use case that requires the rename?

**Comment by blackliner** [comment]
Think of it as a generic docker image that "bazel build ..."'s anything it has in its /src folder? And does tests, and coverage reports, and stuff :D

**Comment by janakdr** [comment]
ok, that's fair. I'm still not convinced that we need to support bazel running in the same logical folder with different names. You can always run `bazel clean` before creating the image, or run with --symlink_prefix=/ to disable symlink creation.

**Comment by blackliner** [comment]
The problem, as stated in my first post, is, that a bazel clean will not work, because it looks for bazel-src symlinks, when it should actually delete bazel-my-project.

**Comment by janakdr** [comment]
I saw that, but are you saying that it won't work on the docker image, or
it won't work in the original workspace? I'm suggesting doing it in the
original workspace before you create the image. Perhaps I'm
misunderstanding something, since I'm not very familiar with docker.

**Comment by blackliner** [comment]
Ok, then try that: go into any bazel workspace, run any bazel command that creates those symlinks. Now rename the folder you are in to something else, and bazel will just stop working. bazel clean will not work, you will need to remove all those symlinks manually.

**Comment by janakdr** [comment]
Right, I agree that once you've renamed the folder, bazel's state is
corrupted. My question is whether you can run `bazel clean` before renaming
the folder, so that the next user of that folder can run bazel from scratch
there. Or is the goal to use existing bazel state with the renamed folder?

**Comment by blackliner** [comment]
It just happens so often during development, that whenever you switch from "local" usage to "docker" usage, bazel throws those random errors. Then you can of course switch back and do a clean there, but it would just be so much more convinient if that all just worked out of the box...

**Comment by janakdr** [comment]
I see. So it's an annoyance in the normal developer workflow to have to run
`bazel clean` every time you update the docker image. That makes sense.
Because I'm ignorant of this, can you tell me what the symlinks point to in
the docker image? Do you also ship bazel's output directory, or are the
symlinks dangling? It seems like they're not dangling because bazel finds
files underneath it?

**Comment by blackliner** [comment]
No worries :D yes its alot about convinience I have to admit. The symlinks do still point to some valid locations, and bazel tries to figure out what to do with all those BUILD and bzl files it finds there. I do not ship anything yet, currently it is all just part of the CI pipeline (or not, yet at least 😛 )

**Comment by janakdr** [comment]
So is the symlink pointing to an output root that bazel created? If so,
does that mean you're putting bazel's output tree in the docker image? That
seems expensive to me, and non-hermetic, since it'll depend on what you
built with bazel previously. Maybe it doesn't matter for the developer
workflow? I could imagine that what you really want is a way to exclude
these symlinks and other incidental files from cluttering up the docker
image in the first place?

**Comment by blackliner** [comment]
I think what happens is that since i also mount the users home directory into the docker, the symlinks get resolved?

**Comment by janakdr** [comment]
Is that desirable? Wouldn't you want to exclude .cache/bazel? Seems
problematic to include all that cruft.

**Comment by blackliner** [comment]
In a perfect world i would want the caching from the host to be reused inside docker, i guess. But yes, leaving out this directory would invalidate the symlinks and thus prevent bazel from going crazy.

**Comment by janakdr** [comment]
Ok, thanks. Given the issue with `bazel clean` that I highlighted earlier,
I think that our first effort would be to enable `bazel clean` to
aggressively delete all symlinks into the output base that start with
`bazel-`, even if they do not match the expected suffix. That would fix
your immediate pain point, I think, and then open up the option for us to
change bazel's behavior to actually match the documentation, although I
could see us going either way on that (it's possible that some users have
workflows built around the directory name at this point). I filed #11980
for the behavior change.  <https://github.com/bazelbuild/bazel/issues/11980>I
don't know how quickly we'll be able to get this done, though. If you think
that you or someone else would be interested in contributing to Bazel by
fixing this, that would be great. It would also help for you or other
people to let us know how big a pain point this is, so we can understand
how to prioritize properly.

**Comment by crt-31** [comment]
This is definitely a pain point. 
My strong preference is for the symlink to be a predictable bazel-<workspace-name> (whatever I put in the WORKSPACE file as workspace(name="myname"). This allows us to use it in things like .bazelignore, and other scripts or tool settings as needed. 

When it is the dir name, it changes based on wherever the dev puts it, and therefore not static and can't be put inside files.
